pkg/registry/apis/provisioning/jobs/driver.go

```diff
@@ -114,27 +126,39 @@ func (d *jobDriver) Run(ctx context.Context) {
 				logger.Error("failed to cleanup jobs", "error", err)
 			}
 		case <-jobTicker.C:
-			d.startDriving(ctx)
+			d.drive(ctx)
 		case <-d.store.InsertNotifications():
-			d.startDriving(ctx)
+			d.drive(ctx)
 		}
 	}
 }
 
-func (d *jobDriver) startDriving(ctx context.Context) {
-	timeoutCtx, cancel := context.WithTimeout(ctx, d.timeout)
-	defer cancel()
-	for timeoutCtx.Err() == nil {
-		if err := d.drive(timeoutCtx); err != nil {
-			if !errors.Is(err, context.Canceled) && !errors.Is(err, ErrNoJobs) {
+// The timing logic here is a bit odd
+// A new thread will be started:
+//   - anytime a new job is added
+//   - periodically jobInterval (eg 30s)
+//
+// Each thread (drive) runs until there are no jobs or has a system error
+func (d *jobDriver) drive(ctx context.Context) {
```

Comment:
Maybe a rename would also help with this one:
```suggestion
func (d *jobDriver) claimAndProcessJobsTillNextInterval(ctx context.Context) {
```